Fix RPC and ProcessGroup GIL deadlock

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45088 Fix RPC and ProcessGroup GIL deadlock**

Fixes #45082

Found a few problems while working on #44983

1. We deliberately swallow RPC timeouts during shutdown, as we haven't
found a good way to handle those. When we convert `_wait_all_workers`
into `_all_gather`, the same logic was inherited. However, as
`_all_gather` meant to be used in more general scenarios, we should
no longer keep silent about errors. This commit let the error throw
in `_all_gather` and also let `shutdown()` to catch them and log.
2. After fixing (1), I found that `UnpickledPythonCall` needs to
acquire GIL on destruction, and this can lead to deadlock when used
in conjuction with `ProcessGroup`. Because `ProcessGroup` ctor is a
synchronization point which holds GIL. In `init_rpc`, followers
(`rank != 0`) can exit before the leader (`rank == 0`). If the two
happens together, we could get a) on a follower, it exits `init_rpc`
after running `_broadcast_to_followers` and before the reaching dtor
of `UnpickledPythonCall`. Then it runs the ctor of `ProcessGroup`,
which holds the GIL and wait for the leader to join. However, the
leader is waiting for the response from `_broadcast_to_followers`,
which is blocked by the dtor of `UnpickledPythonCall`. And hence
the deadlock. This commit drops the GIL in `ProcessGroup` ctor.
3. After fixing (2), I found that `TensorPipe` backend
nondeterministically fails with `test_local_shutdown`, due to a
similar reason as (2), but this time it is that `shutdown()` on a
follower runs before the leader finishes `init_rpc`. This commit
adds a join for `TensorPipe` backend `init_rpc` after `_all_gather`.

The 3rd one should be able to solve the 2nd one as well. But since
I didn't see a reason to hold GIL during `ProcessGroup` ctor, I
made that change too.

Differential Revision: [D23825592](https://our.internmc.facebook.com/intern/diff/D23825592)

mrshenli committed:
Fix RPC and ProcessGroup GIL deadlock

Fixes #45082

Found a few problems while working on #44983

1. We deliberately swallow RPC timeouts during shutdown, as we haven't
found a good way to handle those. When we convert `_wait_all_workers`
into `_all_gather`, the same logic was inherited. However, as
`_all_gather` meant to be used in more general scenarios, we should
no longer keep silent about errors. This commit let the error throw
in `_all_gather` and also let `shutdown()` to catch them and log.
2. After fixing (1), I found that `UnpickledPythonCall` needs to
acquire GIL on destruction, and this can lead to deadlock when used
in conjuction with `ProcessGroup`. Because `ProcessGroup` ctor is a
synchronization point which holds GIL. In `init_rpc`, followers
(`rank != 0`) can exit before the leader (`rank == 0`). If the two
happens together, we could get a) on a follower, it exits `init_rpc`
after running `_broadcast_to_followers` and before the reaching dtor
of `UnpickledPythonCall`. Then it runs the ctor of `ProcessGroup`,
which holds the GIL and wait for the leader to join. However, the
leader is waiting for the response from `_broadcast_to_followers`,
which is blocked by the dtor of `UnpickledPythonCall`. And hence
the deadlock. This commit drops the GIL in `ProcessGroup` ctor.
3. After fixing (2), I found that `TensorPipe` backend
nondeterministically fails with `test_local_shutdown`, due to a
similar reason as (2), but this time it is that `shutdown()` on a
follower runs before the leader finishes `init_rpc`. This commit
adds a join for `TensorPipe` backend `init_rpc` after `_all_gather`.

The 3rd one should be able to solve the 2nd one as well. But since
I didn't see a reason to hold GIL during `ProcessGroup` ctor, I
made that change too.

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -650,7 +650,8 @@ They are used in specifying strategies for reduction collectives, e.g.,
            int,
            int,
-           ::c10d::ProcessGroupGloo::Options>())
+           ::c10d::ProcessGroupGloo::Options>(),
+           py::call_guard<py::gil_scoped_release>())
```
Shouldn't we do this for ProcessGroupNCCL and ProcessGroupMPI as well?

pritamdamania87 commented on torch/distributed/rpc/api.py:
```diff
@@ -197,14 +197,8 @@ def _all_gather(obj):
-                        worker_name=follower_name, timeout=timeout
-                    )
-                )
+            fut.wait()
```
Previously we would call `wait()` on all futures during shutdown, but now we throw on the first one and probably don't wait for the other futures.

It would be nice to keep the same behavior as before. Can we add a try-catch here and save the exception? We can probably only throw the first/last exception we see and catch that in `_wait_all_workers`.

pritamdamania87 reviewed (commented):
Is it possible to add a unit test that would've caught this issue? Even a unit test that would fail once in 100 without this fix would be useful to catch any regressions we have in the future.

rohan-varma commented:
> Is it possible to add a unit test that would've caught this issue? Even a unit test that would fail once in 100 without this fix would be useful to catch any regressions we have in the future.

I think the test described here: https://github.com/pytorch/pytorch/issues/45089 should reproduce the RPC/ProcessGroup GIL deadlock issue, so this should probably be able to use as a unittest for (2).

mrshenli commented:
> Is it possible to add a unit test that would've caught this issue? Even a unit test that would fail once in 100 without this fix would be useful to catch any regressions we have in the future.

After fixing (1), there are multiple existing tests failed due to (2) and (3). This problem didn't surface because existing code swallowed the failure, timed out in 5s, and logged the error message.

mrshenli commented:
> I think the test described here: #45089 should reproduce the RPC/ProcessGroup GIL deadlock issue, so this should probably be able to use as a unittest for (2).

Thanks, let me pull that in as a test.

mrshenli committed:
Update on "Fix RPC and ProcessGroup GIL deadlock"


Fixes #45082

Found a few problems while working on #44983

1. We deliberately swallow RPC timeouts during shutdown, as we haven't
found a good way to handle those. When we convert `_wait_all_workers`
into `_all_gather`, the same logic was inherited. However, as
`_all_gather` meant to be used in more general scenarios, we should
no longer keep silent about errors. This commit let the error throw
in `_all_gather` and also let `shutdown()` to catch them and log.
2. After fixing (1), I found that `UnpickledPythonCall` needs to
acquire GIL on destruction, and this can lead to deadlock when used
in conjuction with `ProcessGroup`. Because `ProcessGroup` ctor is a
synchronization point which holds GIL. In `init_rpc`, followers
(`rank != 0`) can exit before the leader (`rank == 0`). If the two
happens together, we could get a) on a follower, it exits `init_rpc`
after running `_broadcast_to_followers` and before the reaching dtor
of `UnpickledPythonCall`. Then it runs the ctor of `ProcessGroup`,
which holds the GIL and wait for the leader to join. However, the
leader is waiting for the response from `_broadcast_to_followers`,
which is blocked by the dtor of `UnpickledPythonCall`. And hence
the deadlock. This commit drops the GIL in `ProcessGroup` ctor.
3. After fixing (2), I found that `TensorPipe` backend
nondeterministically fails with `test_local_shutdown`, due to a
similar reason as (2), but this time it is that `shutdown()` on a
follower runs before the leader finishes `init_rpc`. This commit
adds a join for `TensorPipe` backend `init_rpc` after `_all_gather`.

The 3rd one should be able to solve the 2nd one as well. But since
I didn't see a reason to hold GIL during `ProcessGroup` ctor, I
made that change too.

Differential Revision: [D23825592](https://our.internmc.facebook.com/intern/diff/D23825592)

[ghstack-poisoned]

mrshenli committed:
Update on "Fix RPC and ProcessGroup GIL deadlock"


Fixes #45082

Found a few problems while working on #44983

1. We deliberately swallow RPC timeouts during shutdown, as we haven't
found a good way to handle those. When we convert `_wait_all_workers`
into `_all_gather`, the same logic was inherited. However, as
`_all_gather` meant to be used in more general scenarios, we should
no longer keep silent about errors. This commit let the error throw
in `_all_gather` and also let `shutdown()` to catch them and log.
2. After fixing (1), I found that `UnpickledPythonCall` needs to
acquire GIL on destruction, and this can lead to deadlock when used
in conjuction with `ProcessGroup`. Because `ProcessGroup` ctor is a
synchronization point which holds GIL. In `init_rpc`, followers
(`rank != 0`) can exit before the leader (`rank == 0`). If the two
happens together, we could get a) on a follower, it exits `init_rpc`
after running `_broadcast_to_followers` and before the reaching dtor
of `UnpickledPythonCall`. Then it runs the ctor of `ProcessGroup`,
which holds the GIL and wait for the leader to join. However, the
leader is waiting for the response from `_broadcast_to_followers`,
which is blocked by the dtor of `UnpickledPythonCall`. And hence
the deadlock. This commit drops the GIL in `ProcessGroup` ctor.
3. After fixing (2), I found that `TensorPipe` backend
nondeterministically fails with `test_local_shutdown`, due to a
similar reason as (2), but this time it is that `shutdown()` on a
follower runs before the leader finishes `init_rpc`. This commit
adds a join for `TensorPipe` backend `init_rpc` after `_all_gather`.

The 3rd one should be able to solve the 2nd one as well. But since
I didn't see a reason to hold GIL during `ProcessGroup` ctor, I
made that change too.

Differential Revision: [D23825592](https://our.internmc.facebook.com/intern/diff/D23825592)

[ghstack-poisoned]

mrshenli commented:
Failures in [MacOS tests](https://circleci.com/api/v1.1/project/github/pytorch/pytorch/7650094/output/105/0?file=true&allocation-id=5f69550c92c4e5711a776ec8-0-build%2F4C621FCC) are already there on Master, see:

#45105 #40434 #40378

lw commented:
The three known failures you linked to were only affecting the ProcessGroup agent. After this diff, those same tests have also started failing for the TensorPipe agent, with the same logs. I opened #45116, #45117 and #45118 to disable them.

mrshenli commented:
> The three known failures you linked to were only affecting the ProcessGroup agent. After this diff, those same tests have also started failing for the TensorPipe agent, with the same logs. I opened #45116, #45117 and #45118 to disable them.

It's actualy failing for both ProcessGroup and TensorPipe, but we don't have an issue to track. See the test below. Thanks for opening those issues. 

https://app.circleci.com/pipelines/github/pytorch/pytorch/215941/workflows/f4bc8989-0016-438b-b029-fa872c487791/jobs/7586991/steps

mrshenli commented:
Trying a fix for those 6 issues: #45126

lw commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -254,6 +254,7 @@ def _tensorpipe_init_backend_handler(store, name, rank, world_size, rpc_backend_
 
     try:
         _tensorpipe_check_device_maps(agent, rpc_backend_options.device_maps)
+        agent.join()
```
This makes me a bit sad, as I thought in principle we wanted to eventually get rid of the `sync` and `join` methods of the agent, and adding more usages will just make it harder. What is the reason for doing this actually? By this time there shouldn't be any RPCs in flight so joining seems to be unnecessary. The only "side effect" I see is that it adds a barrier between all agents. If that's why we use it, can we just make an explicit barrier instead of using join?

mrshenli replied:
> This makes me a bit sad, as I thought in principle we wanted to eventually get rid of the sync and join methods of the agent, and adding more usages will just make it harder. 

Agree, I am sad too. 

> What is the reason for doing this actually? By this time there shouldn't be any RPCs in flight so joining seems to be unnecessary. 

There can be RPCs at this point. Assume there are two processes X and Y. Y could finish `init_rpc` earlier than X and then started sending RPCs to X. At this point, without this join, it is possible that X has not configured the reverse device map, and hence hit failures.

> The only "side effect" I see is that it adds a barrier between all agents. If that's why we use it, can we just make an explicit barrier instead of using join?

Which `barrier` are you referring to? The `dist.barrier()`? Will that add another dependency on `c10d`? I cannot tell which one is worse. :)

lw replied:
I see, thanks for clarifying that there could be RPCs already in flight, I hadn't realized.

So the problem, at a high level, seems to me that we run in troubles if one rank exits init_rpc early and starts doing stuff while other ranks are potentially still inside init_rpc, right? If that's the case, can't we fix it by doing a barrier at the very end of init_rpc?

On one hand, that would nicely match what @pritamdamania is doing for DDP in #45181 (adding a barrier at the end of init_process_group and new_group). Also, since a barrier is a "weaker" operation than joining the agent (because the latter performs multiple barriers), I'd consider a barrier to be more readable as it makes it clear what the core issue is and what the essential solution is, without going "overkill".

And, if we're doing a barrier, we could use _all_gather and thus avoid using both the (deprecated?) agent.join() and without adding a dependency on c10d.

mrshenli replied:
Yep, I would expect `_all_gather` to work too. I actually have one PR for that #44990. Let me land that one too.

mrshenli replied:
Using `_all_gather` makes `test_local_shutdown` flaky. Will come back to debug this after branch cut.

lw replied:
Sounds good, thanks!
